[runtime env] Test common failure scenarios

architkulkarni opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
Tests the following failure scenarios:
- Fail to upload data in `ray.init()` (`working_dir`, `py_modules`)
- Eager install fails in `ray.init()` for some other reason (bad `pip` package)
- Fail to download data from GCS (`working_dir`)

Improves the following error message cases:
- Return RuntimeEnvSetupError on failure to upload working_dir or py_modules
- Return RuntimeEnvSetupError on failure to download files from GCS during runtime env setup

Not covered in this PR:
- RPC to agent fails (This is extremely rare because the Raylet and agent are on the same node.)
- Agent is not started or dead (We don't need to worry about this because the Raylet fate shares with the agent.)

The approach is to use environment variables to induce failures in various places.  The alternative would be to refactor the packaging code to use dependency injection for the Internal KV client so that we can pass in a fake. I'm not sure how much of an improvement this would be.  I think we'd still have to set an environment variable to pass in the fake client, because these are essentially e2e tests of `ray.init()` and we don't have an API to pass it in.
<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

architkulkarni commented:
@rkooo567 Any comments?

rkooo567 commented:
@architkulkarni sorry haven't seen this cuz I deprioritized reviewing PRs lately due to deadlines... I will take a look at it soon

rkooo567 commented on python/ray/tests/test_runtime_env_failure.py:
```diff
@@ -0,0 +1,155 @@
+    del os.environ["_RAY_AGENT_FAILING"]
+
+
+# TODO(SongGuyang): Fail the agent which is in different node from driver.
```
is this still relevant?

architkulkarni replied:
This was an existing test that was simply moved to a new file in this PR.  According to https://github.com/ray-project/ray/pull/23024/files#r824973037 we can remove this test.  I will remove it in this PR, @SongGuyang any concerns with removing it?

rkooo567 commented on python/ray/tests/test_runtime_env_failure.py:
```diff
@@ -0,0 +1,155 @@
+
+# TODO(SongGuyang): Fail the agent which is in different node from driver.
+@pytest.mark.skip(
+    reason="Agent failure will lead to raylet failure and driver failure."
```
remove this?

rkooo567 commented on python/ray/tests/test_runtime_env_failure.py:
```diff
@@ -0,0 +1,155 @@
+    """
+    Test task raises an exception.
+    """
+    with pytest.raises(ray.exceptions.LocalRayletDiedError):
```
This exception doesn't make sense?

rkooo567 commented on python/ray/tests/test_runtime_env_failure.py:
```diff
@@ -0,0 +1,155 @@
+    reason="Agent failure will lead to raylet failure and driver failure."
+)
+@pytest.mark.parametrize("runtime_env_class", [dict, RuntimeEnv])
+def test_runtime_env_broken(
```
Is this the correct name of the test? why is `env_vars={"TF_WARNINGS": "none"}` invalid?

rkooo567 commented on python/ray/tests/test_runtime_env_failure.py:
```diff
@@ -0,0 +1,155 @@
+        else:
+            runtime_env = {"py_modules": [str(tmpdir)]}
+
+        with pytest.raises(Exception):
```
what's the exception that's raised?

rkooo567 commented on python/ray/tests/test_runtime_env_failure.py:
```diff
@@ -0,0 +1,155 @@
+            def f():
+                pass
+
+            with pytest.raises(Exception):
```
Same question as above

rkooo567 commented on python/ray/tests/test_runtime_env_failure.py:
```diff
@@ -0,0 +1,155 @@
+            def f():
+                pass
+
+            with pytest.raises(Exception):
```
I think this is suppose to be `RuntimeEnvSetupError`. Can you catch the precise exceptions here?

rkooo567 reviewed (commented):
> The alternative would be to refactor the packaging code to use dependency injection for the Internal KV client so that we can pass in a fake

I like this approach. How much efforts do you think it'd be?

architkulkarni commented:
> > The alternative would be to refactor the packaging code to use dependency injection for the Internal KV client so that we can pass in a fake
> 
> I like this approach. How much efforts do you think it'd be?

Maybe 0.5-1 weeks?  Maybe we can use something like the Serve KVStore wrapper: https://github.com/ray-project/ray/blob/5c09617471a8c50bd8fe9f85dfda4b0d211d3dbe/python/ray/serve/_private/storage/kv_store.py#L23. As for how to decide at runtime whether to load the real class or the fake, I’m not sure the best way to do it.  The only method that comes to mind is to use an environment variable but I must be missing some standard approach, do you know the right way to do it?

architkulkarni commented on python/ray/_private/runtime_env/packaging.py:
```diff
@@ -592,7 +606,13 @@ async def download_and_unpack_package(
                 if code is None:
-                    raise IOError(f"Failed to fetch URI {pkg_uri} from GCS.")
+                    raise IOError(
+                        f"Failed to download runtime_env file package {pkg_uri} "
```
I think the way would be to examine the logs, `gcs_server.out` has the logs of the form `[runtime env manager] ID to URIs table:
[runtime env manager] URIs reference table:`. I'm not sure how useful it would be for the user to check these logs; this is more likely to be an internal error than something the user can fix.  After https://github.com/ray-project/ray/pull/24719, we haven't heard any reports of this error (but there hasn't been a stable Ray release since then, so we can't be sure that it's been fixed yet)

architkulkarni commented:
@rkooo567 I've addressed the comments.  Thanks for insisting on more precise exceptions, I've added these in the code and in the test and I think it's a good improvement.  They should all return `RuntimeEnvSetupError` now, with the exception of Ray Client `ray.init()` which returns `ConnectionAbortedError` for all server-side failures (though the root cause of the runtime env setup failure still appears in the error message, as shown in the test.)

architkulkarni commented:
`tests:test_state_api` and `test_metrics` were flaky on master, and are unrelated
